README.md
# e-ngx-table

基于Angular的表格组件，耦合后台，仅限公司内部使用。

## Introduce

1. 必须与后台配合，根据配置项自动获取数据

2. 支持分页

3. 支持单元格自定义格式及点击操作

4. 支持高级搜索

5. 支持批量选择行数据

## Usage

1. Install

	```shell
	npm install --save e-ngx-table@latest
	```

2. Set in the .angular-cli.json（@angular/cli）

    ```json
    "styles": [
        "../node_modules/bootstrap/dist/css/bootstrap.min.css",
        "../node_modules/font-awesome/css/font-awesome.min.css"
    ]
    ```

3. Add the ENgxTableModule

	```typescript
	import {ENgxTableModule} from "e-ngx-table";
	@NgModule({
	    imports: [
	        ENgxTableModule
	    ]
	})
	```

4. Use in Template

    ```html
    <e-ngx-table [option]="option" (ready)="ready($event)" (rowSelect)="onRowSelect($event)"></e-ngx-table>
    ```

5. Use in Component

    ```typescript
    table: ENgxTableComponent;

    option: any = {
        serverParam: {
            serverUrl: 'http://192.168.0.88/drainage/TSewerageUserController/getSewerageUserListPage',
            token: '' // 传入登录之后得到的令牌
        },
        operateBtn: [{
                 text: '添加',
                 cls: 'btn-success',
                 event: () => {
                     console.log('添加操作！');
                 }
             }],
        showGlobalSearch: true, // 是否显示全局搜索，默认显示
        showComplexSearch: true, // 是否显示复杂搜索框，默认隐藏
        columns: {
            primaryKey: 'id', // （一般要配置，如果错了rowSelect事件会失效）
            items: [{
                label: "工程名称",
                colName: "name",
                render: (value: any, obj: any) => {
                    return `<span style="color: royalblue;"><span class="glyphicon glyphicon-user"></span>value</span>`;
                },
                event: (data) => {
                    console.log(data);
                }
            }, {
                label: "工程编号",
                colName: "no",
                style: {color: 'red', 'font-weight': 'bold'}
            }, {
                label: "工程地址",
                colName: "address",
                ellipsis: true,
                width: 200
            }, {
                label: "创建时间",
                colName: "createtime",
                search: false,
                complexSearch: true,
                type: 'time'
            }, {
                label: "操作",
                print: false,
                order: null,
                render: [
                    {
                        text: '添加',
                        cls: 'btn-info btn-xs',
                        exist:(obj: any) => { // 根据条件判断是否隐藏按钮
                            return false;
                        },
                        event: (obj: any) => {
                            console.log(obj);
                        }
                    },
                    {
                        text: '编辑',
                        cls: 'btn-info btn-xs',
                        event: (obj: any) => {
                            console.log(obj);
                        }
                    },
                    {
                        text: '刪除',
                        cls: 'btn-danger btn-xs',
                        event: (obj: any) => {
                            console.log(obj);
                        }
                    }
                ]
            }]
        }
    };

    // 表格初始完成
    ready($event) {
        this.table = $event;
        console.log('table ready!');
    }

    // 行选择事件
    onRowSelect($event) {
        console.log($event);
    }

    // 刷新表格
    refresh() {
        this.table.refresh();
    }
    ```

## API

### Inputs

- `option` (`Object`) - 表格配置项，默认配置如下：

```json

serverParam: {
    serverUrl: "", // 服务地址
    token: "", // 令牌
    currentPage: 1, // 当前页
    pageSize: 10, // 每页显示页数
    conditions: [{fieldName: '', value: '', operator: 'LIKE'}], // 查询条件
    orders: [{fieldName: '', value: '', operator: 'DESC'}], // 排序条件
    search: "", // 全局搜索值
    fileds: [] // 全局搜索对应字段
},
operateBtn [] ： 按钮对象数组，对象属性{text: 按钮文本, cls：按钮样式类, event：按钮点击事件
showGlobalSearch: true, // 是否显示全局搜索，默认是true
showComplexSearch: true, // 是否显示复杂搜索框，默认是false，如果设置为true，则在想要进行复杂筛选的items对象里面加两个属性：complexSearch: true,type: 'string','number
columns: {
    primaryKey: "id", // 主键
    batch: true, // 批量选择
    index: { // 序号列
        enabled: true, // 是否启用
        print: true // 是否可以打印
    },
    items: [
        {
            label: "", // 表头标签
            colName: "", // 字段名
            visible: true, // 是否可见
            print: true, // 是否可以打印
            order: 'normal', // 排序，可取值：null, normal, asc, desc
            search: true, // 是否加入全局搜索，默认是true
            complexSearch: {
                enable: false, // 是否能进行复杂筛选，如果不配置complexSearch对象，则默认是false
                type: 'string', // 'string', 'number', 'select'
                data: [{text: '待解决', value: 0}] // 只有type为'select'时，才需要设置该属性，属性值是对象数组
            },
            width: null, // 单元格宽度，如'100px'
            cls: "text-center", // 单元格样式类
            style: null, // 单元格样式
            ellipsis: false, // 文字超出单元格是否显示...
            // 单元格格式化，如果是函数，函数参数(value：当前单元格值, obj：单元格所在行数据对象)，单元格显示函数返回的值。
            // 如果是对象数组，就显示按钮，对象属性{text: 按钮文本, cls：按钮样式类, event：按钮点击事件}
            render: null, //  单元格格式化，如果是函数(value: any, obj: any) => {}，就显示函数返回的值，如果是数组，就显示按钮{text: string 按钮文本, cls: string 按钮样式类, event: function 按钮点击事件回调函数, exist: function 判断按钮是否隐藏函数}
            event: null, // 单元格点击事件, 返回当前行的数据对象
        },
        { // 操作按钮列的设置
            label: "操作",
            print: false,
            order: null,
            exist: (obj) => {
                // 控制按钮的显隐藏，如果返回值为true，则显示该按钮，如果返回值为false，则隐藏该按钮
            },
            render: [
               {
                   text: '编辑',
                   cls: 'btn-info btn-xs',
                   event: (obj: any) => {
                       console.log(obj);
                   }
               },
               {
                   text: '刪除',
                   cls: 'btn-danger btn-xs',
                   event: (obj: any) => {
                       console.log(obj);
                   }
               }
            ]
       }
    ]
}
```

```json
operator的值可以有:
"EQ"; // 相等
"NE"; // 不相等
"LIKE"; // 包含
"GT"; // 大于
"LT"; // 小于
"GTE"; // 大于等于
"LTE"; // 小于等于
"AND"; // 交集，复杂条件使用
"OR"; // 并集，复杂条件使用
"IN"; // IN
"ASC"; // 正序，排序使用
"DESC"; // 倒序，排序使用
```

### Outputs (event)

- `ready` - 表格准备就绪后会触发该事件，参数$event为ENgxTableComponent的实例

- `rowSelect` - 行选择事件，参数$event表示已选行的数据，数据类型为对象数组

- `tableRefresh` - 表格刷新后会触发该事件，参数$event为ENgxTableComponent的实例

### Instance Method

- `refresh(): void` - 刷新列表数据

## Develop

	```shell
	npm install // 安装依赖包

	npm start // 启动项目
	```

# License

[![MIT License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](/LICENSE)


CHANGELOG.md
# 5.2.3（2018-11-20）

- 修复操作列按钮显示逻辑的bug

# 5.2.2（2018-11-20）

- 修复操作列按钮显示逻辑的bug

# 5.2.1（2018-11-05）

- 修复表格渲染的bug

# 5.2.0（2018-11-02）

- 优化表格渲染效率，将渲染函数执行次数优化到最少

# 5.1.1（2018-10-26）

- 默认每列不可排序

# 5.1.0（2018-10-15）

- 修复bug

# 5.0.3（2018-09-29）

- 修复bug

# 5.0.2（2018-05-07）

- 修改说明

# 5.0.1（2018-04-24）

- 升级e-ngx-pagination

# 5.0.0（2018-04-24）

- 升级angular到5.0.0

# 2.4.3（2018-04-03）

- 优化复杂筛选功能

# 2.4.2（2018-03-29）

- 添加复杂筛选功能，操作按钮可以控制显隐藏，全局搜索框可以控制显隐藏

# 2.4.1（2018-02-05）

- 删掉依赖的e-ngx-datetimepicker

# 2.4.0（2017-12-13）

- 将essence-ng2-table更名为e-ngx-table

# 2.1.4（2017-10-30）

- 重构 HttpModule修改为HttpClientModule

# 2.1.3（2017-10-17）

- 无

# 2.1.2（2017-10-17）

- 重构 数据加载时显示加载动画

- 新增 表格刷新事件tableRefresh

# 2.1.1（2017-09-19）

- 新增 单元格点击事件

- 重构 禁用表格底部的列筛选及默认全部字段加入全局筛选

# 2.1.0（2017-09-18）

- 新增 自定义操作按钮

# 2.0.9（2017-09-14）

- 新增 单元格内容过长，则超出的部分省略号显示

# 2.0.8（2017-08-30）

- 新增 readme文件新增对conditions，orders的详细说明

# 2.0.7（2017-08-02）

- 新增 请求header添加URMS_LOGIN_TOKEN

# 2.0.6（2017-06-23）

- 重构 默认不显示批量操作复选框

# 2.0.5（2017-06-22）

- 修复 序号列序号错误的bug

# 2.0.4（2017-06-22）

- 新增 全局搜索输入完成自动查询

# 2.0.3（2017-06-22）

- 新增 引入essence-ng2-datetimepicker方便选择时间

# 2.0.2（2017-06-22）

- 修复 操作列单元格按钮紧贴bug，设置操作列单元格按钮的margin，值为2px

# 2.0.1（2017-06-22）

- 新增 rowSelect - 行选择事件，参数$event表示已选行的数据，数据类型为对象数组

# 2.0.0 (2017-06-19)

## Code Refactoring

- 由于后台接口变化，重构代码


src/e-ngx-table.component.ts
/**
 * Created by laixiangran on 2017-06-22
 */

import { Component, OnInit, OnDestroy, Input, Output, EventEmitter } from '@angular/core';
import { DomSanitizer } from '@angular/platform-browser';
import { FormControl } from '@angular/forms';
import { Observable } from 'rxjs/Observable';
import { Subscription } from 'rxjs/Subscription';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/debounceTime';
import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/operator/switchMap';
import 'rxjs/add/observable/throw';
import * as _ from 'lodash';

import { TableDataModel } from './model/tableDataModel';
import { HttpClient, HttpErrorResponse, HttpHeaders } from '@angular/common/http';

@Component({
	selector: 'e-ngx-table',
	templateUrl: './e-ngx-table.component.html',
	styleUrls: ['./e-ngx-table.component.scss']
})
export class ENgxTableComponent implements OnInit, OnDestroy {
	private getDataSubscription: Subscription;
	private searchInputSubscription: Subscription;

	searchInput: FormControl = new FormControl;

	// 表格是否已经加载完成（第一次）
	tableIsloaded: boolean = false;

	// 用来合并配置的变量
	config: any;

	// 表格数据
	tableData: TableDataModel = null;

	// 搜索值
	searchText: string;

	// 已选择行
	selectedDatas: any[] = [];

	// 搜索/排序中
	isSearching: boolean = false;

	// 复杂筛选列
	complexSearchItems: any[] = [];

	// 复杂筛选条件
	conditionList: any[] = [];

	// 显示高级搜索条件与否
	showComplexSearch: boolean = false;

	// 日期组件配置
	options: any = {
		format: 'YYYY-MM-DD'
	};
	// 传进来的conditions
	tableContions: any;

	/**
	 * 属性设置
	 * @param config
	 */
	@Input()
	set option(config: any) {
		this.config = _.merge({}, this.defaultConfig, config);
		this.tableContions = this.config.serverParam.conditions; // 将传进来的conditons保存起来
		if (this.config.columns.items) {
			let items: any[] = this.config.columns.items.map((item: any) => {
				return _.merge({}, this.defaultItemsConfig, item);
			});
			this.config.columns.items = items;
			this.setServerParam();
			if (this.config.showComplexSearch) {
				this.setComplexSearchItem();
			}
			if (!this.tableIsloaded) {
				this.ready.emit(this);
			} else {
				this.refresh();
			}
		} else {
			throw `columns.items is undefined - 缺少数据列配置，请配置。`;
		}
	}

	/**
	 * 表格初始化完成事件
	 * @type {EventEmitter<any>}
	 */
	@Output()
	private ready: EventEmitter<any> = new EventEmitter<any>(false);

	/**
	 * 行选择事件
	 * @type {EventEmitter<any>}
	 */
	@Output()
	private rowSelect: EventEmitter<any> = new EventEmitter<any>(false);

	/**
	 * 表格刷新完成事件
	 * @type {EventEmitter<any>}
	 */
	@Output()
	private tableRefresh: EventEmitter<any> = new EventEmitter<any>(false);

	// 默认配置参数
	private defaultConfig: any = {
		serverParam: {
			serverUrl: '', // 服务地址
			token: '', // 令牌
			currentPage: 1, // 当前页
			pageSize: 10, // 每页显示页数
			conditions: [], // 查询条件
			orders: [], // 排序条件
			search: '', // 全局搜索值
			fileds: [] // 全局搜索对应字段
		},
		operateBtn: [],
		showGlobalSearch: true,
		showComplexSearch: false,
		columns: {
			primaryKey: 'id', // 主键
			filter: false, // 全列过滤
			batch: false, // 批量选择
			index: { // 序号列
				enabled: true, // 是否启用
				print: true // 是否可以打印
			}
		}
	};

	/**
	 * 数列配置
	 */
	private defaultItemsConfig: any = {
		label: '', // 表头标签
		colName: '', // 字段名
		visible: true, // 是否可见
		print: true, // 是否可以打印
		order: null, // 排序，可取值：null, normal, asc, desc
		search: true, // 是否加入全局搜索
		width: null, // 单元格宽度
		cls: 'text-center', // 单元格样式类
		style: null, // 单元格样式
		ellipsis: false, // 文字超出单元格是否显示...
		complexSearch: {
			enable: false,
			type: 'string',
		}, // 是否可以进行复杂筛选
		type: 'string',
		filterProp: { // 过滤条件
			enabled: false, // 是否启用
			type: 'string', // 字段数据类型，可取值：string, date, select
			operator: 'like', // 操作符号，可取值：like, eq
			value: null // 筛选的值
		},
		render: null, // 单元格格式化，如果是函数(value: any, obj: any) => {}，就显示函数返回的值，如果是数组，就显示按钮{text, cls, event, exist}
		event: null // 单元格点击事件, 返回当前行的数据对象
	};

	constructor(private http: HttpClient, public domSanitizer: DomSanitizer) {}

	ngOnInit() {
		if (this.config.showGlobalSearch) {
			// 订阅全局搜索输入框值变化事件
			this.searchInputSubscription = this.searchInput.valueChanges
				.debounceTime(300) // 延迟300ms
				.distinctUntilChanged() // 输入值没变化，不再发请求
				.switchMap((value: any) => { // 保证请求顺序
					this.isSearching = true;
					this.config.serverParam.search = value;
					return this.getDataServer();
				})
				.subscribe((serverData: any) => {
						if (serverData.code === 'ok') {
							this.handlerTableData(serverData.result);
						} else {
							throw new Error(serverData.info);
						}
					},
					(error: any) => {
						throw new Error(error);
					}
				);
		} else {
			this.getTableData();
		}
	}

	ngOnDestroy() {
		this.tableData = null;
		this.getDataSubscription && this.getDataSubscription.unsubscribe();
		this.searchInputSubscription && this.searchInputSubscription.unsubscribe();
	}

	/**
	 * 高级搜索按钮点击事件
	 */
	searchClick() {
		this.showComplexSearch = !this.showComplexSearch;
	}

	/**
	 * 添加条件按钮点击事件
	 */
	addCondition() {
		this.conditionList.push({
			fieldName: this.complexSearchItems[0].colName,
			operator: 'EQ',
			value: '',
			type: this.complexSearchItems[0].type,
			data: this.complexSearchItems[0].type === 'select' ? this.complexSearchItems[0].data : null
		});
	}

	/**
	 * colName值变化
	 * @param $event
	 * @param item
	 */
	optionChange($event, item) {
		item.fieldName = $event.target.value;
		this.complexSearchItems.forEach((curItem) => {
			if (curItem.colName === $event.target.value) {
				item.type = curItem.type;
				item.data = item.type === 'select' ? curItem.data : null;
			}
		});
	}

	operatorChange($event, item) {
		item.operator = $event.target.value;
	}

	/**
	 * 搜索按钮点击事件
	 */
	search() {
		const conditions = [];
		this.conditionList.forEach((ite) => {
			if (ite.value) {
				conditions.push({fieldName: ite.fieldName, operator: ite.operator, value: ite.value});
			}
		});
		if (this.tableContions.length > 0) {
			this.tableContions.forEach((con) => {
				conditions.push(con);
			})
		}
		this.config.serverParam.conditions = conditions;
		this.getTableData();
	}

	/**
	 * 关闭复杂筛选
	 */
	closeComplexSearch() {
		this.showComplexSearch = false;
		this.conditionList = [];
		const conditions = [];
		if (this.tableContions.length > 0) {
			this.tableContions.forEach((con) => {
				conditions.push(con);
			})
		}
		this.config.serverParam.conditions = conditions;
		this.getTableData();
	}

	/**
	 * 单元格点击事件
	 * @param column
	 * @param data
	 */
	tdClick(column: any, data: any) {
		column.event && column.event(data);
	}

	/**
	 * 按钮点击事件
	 */
	btnClick(btn: any) {
		btn.event && btn.event();
	}

	/**
	 * 获取表格数据
	 */
	getTableData() {
		this.getDataServer().subscribe((serverData: any) => {
			if (serverData.code === 'ok') {
					this.handlerTableData(serverData.result);
				} else {
					throw new Error(serverData.info);
				}
			}, (error: any) => {
				throw new Error(error);
			}
		);
	}

	/**
	 * 处理表格数据
	 */
	handlerTableData(tableData: TableDataModel) {
		tableData.items.forEach((dataItem: any) => {
			dataItem.tdContent = {}; // 每行单元格展示的内容（包括样式）
			dataItem.btnRender = []; // 每行操作列中的按钮组
			this.config.columns.items.forEach((item: any) => {
				dataItem.tdContent[item.colName] = this.domSanitizer.bypassSecurityTrustHtml(dataItem[item.colName]);
				if (item.render && this.isFunction(item.render)) {
					dataItem.tdContent[item.colName] = this.domSanitizer.bypassSecurityTrustHtml(item.render(dataItem[item.colName], dataItem));
				} else if (this.isArray(item.render)) {
					item.render.forEach((render: any) => {
						const temp_render = JSON.parse(JSON.stringify(render));
						temp_render.event = render.event;
						temp_render.isShow = !this.isFunction(render.exist) || render.exist(dataItem);
						dataItem.btnRender.push(temp_render);
					});
				}
			});
		});
		this.tableData = tableData;
		this.isSearching = false;
		if (this.tableIsloaded) {
			this.tableRefresh.emit(this);
		}
		this.tableIsloaded = true;
	}

	/**
	 * 获取表格数据服务
	 * @returns {Observable<any>}
	 */
	getDataServer(): Observable<any> {
		let serverParam: any = this.deepCopyObj(this.config.serverParam);
		delete serverParam.serverUrl;
		delete serverParam.token;
		return this.post(this.config.serverParam.serverUrl, serverParam);
	}

	/**
	 * 设置查询参数
	 */
	setServerParam() {
		this.config.columns.items.forEach((col: any) => {
			this.setOrders(col, true);
			if (col.filterProp && col.filterProp.enabled) {
				if (col.filterProp.type !== 'select' && col.filterProp.value) {
					this.config.serverParam.conditions.push({
						fieldName: col.colName,
						value: col.filterProp.value,
						operator: col.filterProp.operator.toUpperCase()
					});
				}
			}
			if (col.search && col.colName) {
				this.config.serverParam.fileds.push(col.colName);
			}
		});
	}

	/**
	 * 设置复杂筛选的列
	 */
	setComplexSearchItem() {
		this.config.columns.items.forEach((col: any) => {
			if (col.complexSearch.enable && col.colName) {
				if (col.complexSearch.type === 'select') {
					this.complexSearchItems.push({
						label: col.label,
						colName: col.colName,
						type: 'select',
						data: col.complexSearch.data
					})
				} else {
					this.complexSearchItems.push({
						label: col.label,
						colName: col.colName,
						type: col.complexSearch.type
					})
				}
			}
		});
	}

	/**
	 * 设置排序条件
	 * @param column 列对象
	 * @param isFirst 是否第一次设置
	 */
	setOrders(column: any, isFirst: boolean = false): void {
		let orders: any[] = this.config.serverParam.orders;
		if (isFirst) {
			column.operator = column.order;
			if (column.order === 'asc' || column.order === 'desc') {
				orders.push({
					fieldName: column.colName,
					operator: column.operator.toUpperCase()
				});
			}
		} else {
			let existIndex: number = 0;
			let isExist: boolean = orders.some((order: any, index: number) => {
				if (order.fieldName === column.colName) {
					existIndex = index;
				}
				return order.fieldName === column.colName;
			});
			if (isExist) {
				if (column.operator == 'asc') {
					column.operator = 'desc';
					orders[existIndex].operator = column.operator.toUpperCase();
				} else if (column.operator == 'desc') {
					column.operator = 'normal';
					orders.splice(existIndex, 1);
				}
			} else {
				column.operator = 'asc';
				orders.push({
					fieldName: column.colName,
					operator: column.operator.toUpperCase()
				});
			}
		}
	}

	/**
	 * 排序方法
	 * @param column
	 * @private
	 */
	sort(column: any): void {
		if (column.order) {
			this.isSearching = true;
			this.setOrders(column);
			this.refresh();
		}
	}

	/**
	 * 页面切换
	 * @param event
	 */
	pageChanged(event: any): void {
		this.refresh();
	};

	/**
	 * 判断是否为方法
	 * @param param
	 * @returns {boolean}
	 */
	isFunction(param: any): boolean {
		return typeof param === 'function';
	}

	/**
	 * 判断是否为数组
	 * @param param
	 * @returns {boolean}
	 */
	isArray(param: any): boolean {
		return Array.isArray(param);
	}

	/**
	 * 深度拷贝对象
	 * @param obj
	 * @returns {any}
	 */
	deepCopyObj(obj: any): any {
		return JSON.parse(JSON.stringify(obj));
	}

	/**
	 * 复选框状态改变事件
	 * @param data
	 * @private
	 */
	checkboxChangeEvent(data) {
		data.selected = !data.selected;
		let existIndex: number = 0;
		let isExist: boolean = this.selectedDatas.some((selectedData: any, index: number) => {
			if (selectedData[this.config.columns.primaryKey] === data[this.config.columns.primaryKey]) {
				existIndex = index;
			}
			return selectedData[this.config.columns.primaryKey] === data[this.config.columns.primaryKey];
		});
		if (data.selected) {
			if (!isExist) {
				let selectData: any = this.deepCopyObj(data);
				delete selectData.selected;
				this.selectedDatas.push(selectData);
			}
		} else {
			if (isExist) {
				this.selectedDatas.splice(existIndex, 1);
			}
		}
		this.rowSelect.emit(this.selectedDatas);
	}

	/**
	 * 行选择事件
	 * @param data
	 */
	rowChangeEvent(data) {
		if (!this.config.columns.batch) {
			this.selectedDatas = [];
			this.tableData.items.forEach((item: any) => {
				if (data[this.config.columns.primaryKey] === item[this.config.columns.primaryKey]) {
					item.selected = !item.selected;
					if (item.selected) {
						let selectData: any = this.deepCopyObj(item);
						delete selectData.selected;
						this.selectedDatas.push(selectData);
					}
				} else {
					item.selected = false;
				}
			});
			this.rowSelect.emit(this.selectedDatas);
		}
	}

	/**
	 * 全选复选框状态改变事件
	 * @param $event
	 */
	checkboxAllChangeEvent($event) {
		let checked = $event.target.checked;
		for (let i = 0; i < this.tableData.items.length; i++) {
			this.tableData.items[i].selected = checked;
		}
		this.selectedDatas = checked ? this.deepCopyObj(this.tableData.items) : [];
		this.rowSelect.emit(this.selectedDatas);
	}

	/**
	 * 刷新列表数据
	 */
	refresh(): void {
		this.getTableData();
	}

	/**
	 * 优化ngFor
	 * @param index
	 * @param data
	 * @returns {any}
	 */
	trackById(index: any, data: any): any {
		return data[this.config.columns.primaryKey];
	}

	/**
	 * post请求
	 * @param url 请求路径
	 * @param body 请求body
	 * @returns {Observable<any>}
	 */
	post(url: string, body: any = null): Observable<any> {
		let headers: HttpHeaders = new HttpHeaders({
			'Content-Type': 'application/json',
			'Authorization': this.config.serverParam.token,
			'URMS_LOGIN_TOKEN': this.config.serverParam.token
		});
		let options = {
			headers: headers
		};
		return this.http.post(url, body && JSON.stringify(body), options).catch((error: HttpErrorResponse) => {
			let errMsg = (error.message) ? error.message :
				error.status ? `${error.status} - ${error.statusText}` : 'Server error';
			return Observable.throw(errMsg);
		});
	}

	/**
	 * 操作列按钮点击
	 * @param {MouseEvent} evt
	 * @param button
	 * @param data
	 */
	handlerBtnClick(evt: MouseEvent, button: any, data: any) {
		evt.stopPropagation();
		button.event(data);
	}
}
